### PR TITLE
Add in_shipping_zones attribute to shop countries query

### PR DESCRIPTION
**Pull request by jakubkuc**
I want to merge this change because  API query for fetching countries with or without shipping zones was needed.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-5075 Add query to get countries not selected in any shipping zone](https://app.clickup.com/t/2549495/SALEOR-5075)

File: saleor/graphql/shop/types.py [hunk 1]
```diff
@@ -128,6 +129,10 @@ class Shop(graphene.ObjectType):
                 "A language code to return the translation for."
             ),
         ),
+        in_shipping_zones=graphene.Boolean(
+            description="Only countries that have shipping zones assigned",
```
Comment (maarcingebala) [inline saleor/graphql/shop/types.py:133]: I think we can document all cases as each has a different result, for example: "When `true`, return X; when `false`, return Y; when the argument is not provided (`null`), return Z."

File: saleor/graphql/shop/types.py [hunk 2]
```diff
@@ -222,16 +227,16 @@ def resolve_available_shipping_methods(_, info, channel, address=None):
...
         taxes = {vat.country_code: vat for vat in VAT.objects.all()}
-
         # DEPRECATED: translation.override will be dropped in Saleor 4.0
         with translation.override(language_code):
             return [
                 CountryDisplay(
                     code=country[0], country=country[1], vat=taxes.get(country[0])
                 )
                 for country in countries
+                if country[0] in get_countries_codes_list(in_shipping_zones)
```
Comment (maarcingebala) [inline saleor/graphql/shop/types.py:239]: Can't we have something like this? Now we iterate over `django_countries.countries` twice- in the resolver and in the utility function while I think we can do it once.

```suggestion
                for country in get_countries_codes_list(in_shipping_zones)
```

File: saleor/graphql/shop/utils.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from typing import Optional
+
+from django_countries import countries
+
+from ...shipping.models import ShippingZone
+
+
+def get_countries_codes_list(in_shipping_zones: Optional[bool] = None):
+    """Return set of countries codes."""
```
Comment (maarcingebala) [inline saleor/graphql/shop/utils.py:9]: We should also describe the parameter (it could have the same description as in the schema).

**Review by maarcingebala — COMMENTED** [review]
Now we only have a unit test for the utility function (which is good), but we should also have a test that uses the new parameter in some API call. In general, all API fields should be used in some tests; otherwise, we wouldn't know that we accidentally broke a field or resolver, while the utility function would keep passing.

File: saleor/graphql/shop/types.py (hunk 1 above)
Comment (L3str4nge) [inline saleor/graphql/shop/types.py:133]: "." 🚓

File: saleor/graphql/shop/types.py [hunk 4]
```diff
@@ -222,16 +227,16 @@ def resolve_available_shipping_methods(_, info, channel, address=None):
         return resolve_available_shipping_methods(info, channel, address)
 
     @staticmethod
-    def resolve_countries(_, _info, language_code=None):
+    def resolve_countries(_, _info, language_code=None, in_shipping_zones=None):
```
Comment (L3str4nge) [inline saleor/graphql/shop/types.py:230]: Shouldn't be `in_shupping_zones=False` by default?

File: saleor/graphql/shop/utils.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+from typing import Optional
+
+from django_countries import countries
+
+from ...shipping.models import ShippingZone
+
+
+def get_countries_codes_list(in_shipping_zones: Optional[bool] = None):
```
Comment (L3str4nge) [inline saleor/graphql/shop/utils.py:8]: I am not sure `Optional[bool]` is good approach. I think if we want to pass `bool` then it should be True/False by default.

File: saleor/graphql/shop/types.py [hunk 6]
```diff
@@ -128,6 +129,10 @@ class Shop(graphene.ObjectType):
                 "A language code to return the translation for."
             ),
         ),
+        in_shipping_zones=graphene.Boolean(
```
Comment (korycins) [inline saleor/graphql/shop/types.py:132]: Isn't it something that could be placed under `filter` field? At the first moment I wasn't sure what this field exactly does.

File: saleor/graphql/shop/types.py (hunk 4 above)
Comment (jakubkuc) [reply saleor/graphql/shop/types.py:233]: I added better description to this field. It can be None - in this case it returns all of the countries and that is what we want by default.

File: saleor/graphql/shop/types.py (hunk 6 above)
Comment (jakubkuc) [reply saleor/graphql/shop/types.py:132]: I added better description. I hope it will be clear now

File: saleor/graphql/shop/utils.py (hunk 5 above)
Comment (jakubkuc) [reply saleor/graphql/shop/utils.py:8]: I added better description. We want to be able to not pass this argument at all (as we were up till now) - which will cause all countries codes to be returned.

File: saleor/graphql/shop/tests/test_shop_utils.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+
+def test_get_countries_codes_list(shipping_zones):
+    # given
+    fixture_countries_code_set = {zone.countries[0].code for zone in shipping_zones}
+
+    # when
+    countries_list_all = get_countries_codes_list()
+    countries_list_false = get_countries_codes_list(False)
+    countries_list_true = get_countries_codes_list(in_shipping_zones=True)
```
Comment (IKarbowiak) [inline saleor/graphql/shop/tests/test_shop_utils.py:11]: There should be 3 different tests.

File: saleor/graphql/shop/types.py (hunk 2 above)
Comment (jakubkuc) [reply saleor/graphql/shop/types.py]: It will not work because translation.override works only when iteration is made on djangos `Countries` object. I can't see the way to do it with less iterations than 2 at the moment. I might miss something though.

File: saleor/graphql/shop/filters.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+import graphene
+
+
+class CountryFilterInput(graphene.InputObjectType):
+    in_shipping_zones = graphene.Boolean()
```
Comment (maarcingebala) [inline saleor/graphql/shop/filters.py:5]: Missing description.

File: saleor/graphql/shop/filters.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import graphene
+
+
+class CountryFilterInput(graphene.InputObjectType):
+    in_shipping_zones = graphene.Boolean(
```
Comment (korycins) [inline saleor/graphql/shop/filters.py:5]: Wouldn't be better to have something like: `attached_to_shipping_zones`? `in_shipping_zones` for first look sugested me that I can provide IDs of shipping zones. 
Just want to know what do you think

File: saleor/graphql/shop/resolvers.py [hunk 10]
```diff
@@ -25,6 +31,22 @@ def resolve_available_shipping_methods(info, channel_slug: str, address):
     ]
 
 
+def resolve_countries(**kwargs):
+    countries_filter = kwargs.get("filter", {})
+    in_shipping_zones = countries_filter.get("in_shipping_zones")
+    language_code = kwargs.get("language_code")
+    taxes = {vat.country_code: vat for vat in VAT.objects.all()}
+    # DEPRECATED: translation.override will be dropped in Saleor 4.0
```
Comment (korycins) [inline saleor/graphql/shop/resolvers.py:39]: Make sure that you created a task for this and attached it to Saleor4.0 list

File: saleor/graphql/shop/filters.py (hunk 9 above)
Comment (jakubkuc) [reply saleor/graphql/shop/filters.py:5]: now that you mention it I can see that it might be confusing. I'll change it.

**Review by maarcingebala — APPROVED** [review]

File: saleor/graphql/shop/tests/test_shop.py [hunk 11]
```diff
@@ -1334,3 +1334,55 @@ def test_get_shop_limit_info_returns_null_by_default(staff_api_client):
...
+COUNTRY_FILTER_QUERY = """
+    query($filter: CountryFilterInput!) {
+        shop {
+            countries(filter: $filter){
+            code
+            }
+        }
+    }
+
+"""
```
Comment (IKarbowiak) [inline saleor/graphql/shop/tests/test_shop.py:1348]: I will suggest adding two additional tests for empty "filter" option and for `{"attachedToShippingZones": None}`.

**Review by IKarbowiak — APPROVED** [review]

**Review by fowczarek — APPROVED** [review]

**Comment by jakubkuc** [comment]
> Now we only have a unit test for the utility function (which is good), but we should also have a test that uses the new parameter in some API call. In general, all API fields should be used in some tests; otherwise, we wouldn't know that we accidentally broke a field or resolver, while the utility function would keep passing.

Tests added. Will add some more as @IKarbowiak suggested

File: saleor/graphql/shop/tests/test_shop.py (hunk 11 above)
Comment (jakubkuc) [reply saleor/graphql/shop/tests/test_shop.py:1348]: empty filter option is already covered in old tests, the 'None' option I will add
